nncf/common/hook_handle.py

```diff
@@ -0,0 +1,63 @@
+# Copyright (c) 2023 Intel Corporation
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#      http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import weakref
+from typing import Any, Dict, Union
+
+
+class HookHandle:
+    """
+    A handle to remove a hook.
+    """
+
+    def __init__(self, hooks_registry: Dict[Any, Any]):
+        """
+        :param hooks_registry: A dictionary of hooks, indexed by hook `id`.
+        """
+        self.hooks_registry_ref = weakref.ref(hooks_registry)
+        self._hook_id = None
+        self._op_registered = False
+
+    @property
+    def hook_id(self) -> Union[int, str]:
+        if self._hook_id is None:
+            raise RuntimeError("Attempt to use HookHandle hook id before actual hook registration.")
+        return self._hook_id
+
+    def add(self, op: Any) -> None:
```

Comment:
`HookHandle` should not has the add method, because this is unmutable object for the hook. See example: https://github.com/pytorch/pytorch/blob/7dac2f9f2d0e41c2d4b07e4a673d448e63a4a819/torch/utils/hooks.py#L9